include/dlaf/common/pipeline.h

```diff
@@ -32,25 +34,75 @@ class Pipeline {
   using AsyncRwMutex = pika::execution::experimental::async_rw_mutex<T>;
 
 public:
-  using Wrapper = typename AsyncRwMutex::readwrite_access_type;
-  using Sender = pika::execution::experimental::unique_any_sender<Wrapper>;
+  using ReadOnlyWrapper = typename AsyncRwMutex::read_access_type;
+  using ReadWriteWrapper = typename AsyncRwMutex::readwrite_access_type;
+  using ReadOnlySender = pika::execution::experimental::any_sender<ReadOnlyWrapper>;
+  using ReadWriteSender = pika::execution::experimental::unique_any_sender<ReadWriteWrapper>;
 
   /// Create a Pipeline by moving in the resource (it takes the ownership).
   explicit Pipeline(T object) : pipeline(std::move(object)) {}
-  Pipeline(Pipeline&&) = default;
-  Pipeline& operator=(Pipeline&&) = default;
+
+  Pipeline(Pipeline&& other)
+      : pipeline(std::exchange(other.pipeline, std::nullopt)),
+        nested_sender(std::exchange(other.nested_sender, std::nullopt)) {}
+
+  Pipeline& operator=(Pipeline&& other) {
+    if (this != &other) {
+      pipeline = std::exchange(other.pipeline, std::nullopt);
+      nested_sender = std::exchange(other.nested_sender, std::nullopt);
+    }
+
+    return *this;
+  };
+
   Pipeline(const Pipeline&) = delete;
   Pipeline& operator=(const Pipeline&) = delete;
 
+  ~Pipeline() {
+    release_parent_pipeline();
+  }
+
   /// Enqueue for the resource.
   ///
   /// @return a sender that will become ready as soon as the previous user releases the resource.
   /// @pre valid()
-  Sender operator()() {
+  ReadWriteSender operator()() {
```

Comment:
Rename `readwrite`?

Reply:
Renamed in the latest commits. However, I also kept a deprecated `operator()` in https://github.com/eth-cscs/DLA-Future/pull/993/commits/13c1d2a2c40248c8191b6f36d75d6b315f5366de in case it's useful for @albestro's changes. I'll remove it if we don't see a need for it.